// animeflv-api/dist/functions/getAnimeInfo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAnimeInfo = void 0;
const constants_1 = require("../constants");
const cheerio_1 = require("cheerio");
const cloudscraper_1 = __importDefault(require("cloudscraper"));
async function getAnimeInfo(animeId) {
    if (!animeId || (typeof animeId !== "string")) {
        throw new TypeError(`El parámetro animeId debe ser una string no vacía, pasaste: ${animeId}`, { cause: `animeId: ${animeId}` });
    }
    try {
        constants_1.CloudscraperOptions.uri = 'https://www3.animeflv.net/anime/' + animeId;
        const animeData = (await (0, cloudscraper_1.default)(constants_1.CloudscraperOptions));
        const $ = (0, cheerio_1.load)(animeData);
        const animeInfo = {
            title: $('body > div.Wrapper > div > div > div.Ficha.fchlt > div.Container > h1').text(),
            alternative_titles: [],
            status: $('body > div.Wrapper > div > div > div.Container > div > aside > p > span').text(),
            rating: $('#votes_prmd').text(),
            type: $('body > div.Wrapper > div > div > div.Ficha.fchlt > div.Container > span').text(),
            cover: 'https://animeflv.net' + $('body > div.Wrapper > div > div > div.Container > div > aside > div.AnimeCover > div > figure > img').attr('src'),
            synopsis: $('body > div.Wrapper > div > div > div.Container > div > main > section:nth-child(1) > div.Description > p').text(),
            genres: $('body > div.Wrapper > div > div > div.Container > div > main > section:nth-child(1) > nav > a').text().split(/(?=[A-Z])/),
            episodes: [],
            url: constants_1.CloudscraperOptions.uri
        };
        for (let i = 1; i <= JSON.parse($('script').eq(15).text().match(/episodes = (\[\[.*\].*])/)?.[1]).length; i++) {
            animeInfo.episodes.push({
                number: i,
                url: 'https://www3.animeflv.net/ver/' + animeId + '-' + i
            });
        }
        ;
        $('body > div.Wrapper > div > div > div.Ficha.fchlt > div.Container > div:nth-child(3) > span').each((i, el) => {
            animeInfo.alternative_titles.push($(el).text());
        });
        return animeInfo;
    }
    catch {
        return null;
    }
}
exports.getAnimeInfo = getAnimeInfo;
